libraries/chain/block_root_processor.cpp

```diff
@@ -0,0 +1,114 @@
+}
+
+bool block_root_processor::calculate_root_blocks(uint32_t block_num)
+{
+   bool stored = false;
+   root_storage root_transactions = _processor->retrieve_root_transactions(block_num);
+   ilog("calculate_root_blocks block_num: ${bn}, roots size: ${rs}",("bn", block_num)("rs", root_transactions.size()));
+   for(auto& instance : root_transactions) {
+      const auto& contract = instance.first;
+      auto& transactions = instance.second;
+      ilog("calculate_root_blocks contract: ${contract}, size: ${size}",
+           ("contract",contract.first.to_string())("size",transactions.size()));
+      if (transactions.empty()) {
+         continue;
+      }
+      stored = true;
+      auto& contract_root_idx = _db.get_index<contract_root_multi_index, by_contract>();
+      auto itr = contract_root_idx.find(boost::make_tuple(contract.first, contract.second));
+      const auto merkle_root = chain::merkle(transactions);
+      if (itr == contract_root_idx.end()) {
+         ilog("calculate_root_blocks new root");
+         const auto previous_root_id = chain::checksum256_type();
+         const auto curr_root_id = compute_curr_root_id(previous_root_id, merkle_root);
+         _db.create<contract_root_object>([&](auto& obj) {
+            obj.contract = contract.first;
+            obj.root_name = contract.second;
+            obj.block_num = block_num;
+            obj.root_id = curr_root_id;
+            obj.prev_root_id = previous_root_id;
+            obj.prev_root_bn = 0;
+            obj.merkle_root = merkle_root;
+         });
+      } else {
+         ilog("calculate_root_blocks update");
+         const auto previous_root_id = itr->root_id;
+         const auto curr_root_id = compute_curr_root_id(previous_root_id, merkle_root);
+         const auto previous_root_block_number = itr->block_num;
+         _db.modify(*itr, [&](auto& obj) {
+            obj.block_num = block_num;
+            obj.root_id = curr_root_id;
+            obj.prev_root_id = previous_root_id;
+            obj.prev_root_bn = previous_root_block_number;
+            obj.merkle_root = merkle_root;
+         });
+      }
+   }
+   return stored;
+}
+
+uint32_t block_root_processor::extract_root_block_number(const chain::checksum256_type& root_id) {
+   // Extract the pointer to the data from the checksum.
+   auto root_id_data = root_id.data();
+    
+   // Since we are dealing with a checksum256_type, which is typically an array of bytes,
+   // we need to correctly handle the conversion to uint32_t considering endianess.
+   uint32_t root_block_number;
+    
+   // Copy the first 4 bytes of the data into a uint32_t variable.
+   // memcpy ensures that we handle any alignment issues.
+   std::memcpy(&root_block_number, root_id_data, sizeof(uint32_t));
+
+   // Reverse the endianess if necessary.
+   // You might need to adjust this based on how the data is stored (big-endian vs little-endian).
+   return fc::endian_reverse_u32(root_block_number);
```

Comment:
odd comment.  can we not be certain which endian we're dealing with (or use a macro)?